### DISCUSSION TITLE
Missing file "Step_init.c"

### DISCUSSION THREAD

**Discussion by mtraboulsi1** [OpenModelica/OpenModelica#10325] [Q&A] [answered]
I have recently exported the C code for the step block from OpenModelica, but while reviewing the code, I encountered an error due to the absence of the "Step_init.c" file which was referenced in the code via the #include statement. Despite having an alternative file named "Step_init.xml", I cannot use it in the C source code.

I'm curious if there is an additional option in OpenModelica that allows to generate this file "Step_init.c" . If not could you offer some ideas on how to solve this problem.

Thank you in advance!

**Comment by AnHeuermann** [comment]
Hey @mtraboulsi1. Can you tell us a bit more on how to reproduce this? What version of OpenModelica are you using, on what operation system are you, what exactly did you do to encounter this issue?
What exactly is the error message?

-----------

I'm building the `Modelica.Blocks.Sources.Step` from the Modelica 4.0.0 library:

```modelica
loadModel(Modelica, {"4.0.0"}); getErrorString();
buildModel(Modelica.Blocks.Sources.Step); getErrorString();
```

Are you referring to following C code from `Modelica.Blocks.Sources.Step.c?

```C
#if defined(OPENMODELICA_XML_FROM_FILE_AT_RUNTIME)
  [...]
#else
  data->modelData->initXMLData =
  #include "Modelica.Blocks.Sources.Step_init.c"
    ;
#endif
```

In the default simulation case `OPENMODELICA_XML_FROM_FILE_AT_RUNTIME` is defined in the Makefile, so it's no issue that the file is missing.

If we instead export the model as a 1.0 ModelExchange FMU

```modelica
loadModel(Modelica, {"4.0.0"}); getErrorString();
translateModelFMU(Modelica.Blocks.Sources.Step, version="1.0", fmuType="me");
```

the file `Modelica_Blocks_Sources_Step.fmu/sources/Modelica_Blocks_Sources_Step_info.c` is generated and has the XML file `Modelica_Blocks_Sources_Step_info.xml` encoded as one large `const char*` that will be directly compiled into the program.
The reason was probably to not need a full XML parser inside the FMU. But for FMI 2.0 we do it anyway.

**Reply by mtraboulsi1** [reply]
Hello @AnHeuermann I use the same version as you the Modelica 4.0.0
So, I used the Step block and I run the simulation with the C/C++ Compiler Flags (Optional) in order to obtain a C source code. Then, I tested this file with Visual studio code but I got an error concerning #include " Step_init.c" because this file does not exist, I have an equivalent file "Step_init.xml".
The error is "cannot open source file "Step_init.c" and "Step_info.c".

And this is the part of the code that is causing the error :

#if defined(OPENMODELICA_XML_FROM_FILE_AT_RUNTIME)
  data->modelData->initXMLData = NULL;
  data->modelData->modelDataXml.infoXMLData = NULL;
  #else
  #if defined(_MSC_VER) /* handle joke compilers */
  {
  /* for MSVC we encode a string like char x[] = {'a', 'b', 'c', '\0'} */
  /* because the string constant limit is 65535 bytes */
  static const char contents_init[] =
    #include "Step_init.c"
    ;
  static const char contents_info[] =
    #include "Step_info.c"
    ;
    data->modelData->initXMLData = contents_init;
    data->modelData->modelDataXml.infoXMLData = contents_info;
  }
  #else /* handle real compilers */
  data->modelData->initXMLData =
  #include "Step_init.c"
    ;
  data->modelData->modelDataXml.infoXMLData =
  #include "Step_info.c"
    ;
  #endif /* defined(_MSC_VER) */
  #endif /* defined(OPENMODELICA_XML_FROM_FILE_AT_RUNTIME) */

**Reply by AnHeuermann** [reply]
Is there a reason you are using the Visual Studio Compiler? Otherwise I would suggest to use the default MinGW GNU compiler shipped with OpenModelica:

![grafik](https://user-images.githubusercontent.com/38031952/223145068-873f9d25-e97a-468c-86c5-c5867769776e.png)

You can change it in OMEdit in `Tools->Options->Simulation->Target Build`: `MinGW`

If you really need to use MSVC you can define `OPENMODELICA_XML_FROM_FILE_AT_RUNTIME` at compile-time to solve this specific issue. But it is very likely that you'll encounter more issues. The MSVC compiler tool-chain isn't integrated very well in OpenModelica and more or less broken.

**Reply by mtraboulsi1** [reply]
Yes, I'm using MinGW GNU but I want to test the source file c independently on Visual studio code. So, my idea is to make a model on OpenModelica then convert this model into C source code in order to test it independently on Visual studio code.

**Reply by AnHeuermann** [reply]
Sorry, I'm lost. What exactly are you trying to achieve? What are you trying to test?

**Reply by mtraboulsi1** [reply]
I used openmodelica's C/C++ compiler flags (optional)  in (Simulation Setup - Step) to convert the Step block to C source code for testing on Visual studio code. But I can't run this C code because of two missing files.

**Comment by mtraboulsi1** [comment]
@AnHeuermann

![Question](https://user-images.githubusercontent.com/127110869/224027823-9c55c0e0-ff70-4b21-b759-1103bae05cc2.PNG)
 I have the C project which generates with a "#include Step_info.c" directive. While in my project I do not have the “Step_info.c” file but rather the “Step_info. json".

**Reply by AnHeuermann** [reply]
Check the generated Makefile `ModelName.makefile`. It defines `OPENMODELICA_XML_FROM_FILE_AT_RUNTIME`. So just define the variable at compile time.

**Reply by mtraboulsi1** [reply]
@AnHeuermann Thank you for your answer. However, as a new user of OpenModelica, I would like to know if you have any examples for modifying the compilation flags to enable JSON output. Could you provide me with some, please?

**Reply by AnHeuermann** [reply]
I still don't know what you are trying to achieve. Please take the time to provide a step-by-step explanation on what you are trying to do and what your goal is. Why do you need the JSON as a C-string? It is not needed for the compilation process. 

If you are a new to OpenModelica there is probably not much of a point in trying to analyze the generated C code of the compiler.

So to help you you need to provide as much information as possible. I asked a bunch of questions at the beginning of this discussion, but you didn't provide an answer.

  - What version of OpenModelica are you using?
    This is not the version of the Modelica library, that is the language. You can check the version of the compiler in OMEdit: `Help->About OMEdit`
  - What operation system are you using?
    The C compilation depends on the OS, there is no general answer on how to compile the generated C source files.
  - Step-by-step explanation on what you did. Make it reproducible for everyone, than you chances to get help are way higher.
  - What is the overall goal? What is the problem you try to accomplish?
  
-------------------

The general process is that you have a Modelica model. This is not an OpenModelica model. Modelica is the language, OpenModelica is one tool (of many) that can simulate models of the Modelica language.

So for `Step.mo`
```
model Step
 extends Modelica.Blocks.Sources.Step;
end Step;
```

Than you can build a simulation executable from that Modelica model `Step` with OMEdit or by calling the OpenModelica Compiler (OMC) directly. For the purpose of creating reproducible examples we are usually using MOS-scripts which use the [scripting application interface](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/scripting_api.html) to OMC

```
loadFile("Step.mo"); getErrorString();
simulate(Step); getErrorString();
```

Internally the compiler will do a lot of stuff and at the end generates C code for the model and compiles that into a run-able simulation executable `Step` or `Step.exe` if you are on Windows.

While it is possible to examine the generated C code, e.g. `Step.c`, and re-compile the executable, there is usually no reason to do so. (Unless you are trying to find some specific issue.)
Depending on your OS and setup it is not trivial to re-compile the C files.
So in the same way it is not trivial to setup your C/C++ tools to analyze the generated code.

**Comment by mtraboulsi1** [comment, marked as answer]
Hello @AnHeuermann, I attached a PDF file to explain step by step my objective and what I am trying to do.
[Modelica model to C.pdf](https://github.com/OpenModelica/OpenModelica/files/11015916/Modelica.model.to.C.pdf)
 Thanks in advance for your help.

**Reply by AnHeuermann** [reply]
Okay, there seems to be some misunderstandings:

With the compiler flag `C/C++ Compiler Flags (Optional)`, see [--cppFlags](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-cppflags), you can add additional flags to the C compiler. The OpenModelica compiler always generate C code from a Modelica models (unless you explicitly change it to e.g. C++ code, but doesn't matter here) and then compiles it into an executable program. With the flag enabled you can for example change the optimization level uses by the C/C++ compiler and stuff like that.

You can debug the generated executable `Modelica.Blocks.Sources.Step` / `Modelica.Blocks.Sources.Step.exe` in VSCode, but you'll need to use the right tool-chain. If you are on Windows and compiled with gcc or clang form the OMDev environment (MSYS) you should use GDB from OMDev to debug. With MSVC you'll have a very very hard time to get anything compiled.
You can follow https://github.com/OpenModelica/OpenModelica/wiki/Debugging-with-GDB#debug-cc-runtime-and-generated-model-code
If you are on Windows this tutorial assumes you are using OMDev.
You are now very deep inside the guts of the OpenModelica compiler and far away from the "beginner" territory.

Here is a launch.json that I would use to debug the executable in VSCode. You can configure VSCode to do everything the Makefile does and compile the C code, but why? It is harder and doesn't add anything new.

```json
{
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
  "version": "0.2.0",
  "configurations": [
  {
    "name": "(gdb) Step",
    "type": "cppdbg",
    "request": "launch",
    "program": "/path/to/discussion-10325/temp/Modelica.Blocks.Sources.Step",
    "args": [],
    "stopAtEntry": false,
    "cwd": "/path/to/discussion-10325/temp/",
    "environment": [],
    "externalConsole": false,
    "MIMode": "gdb",
    "setupCommands": [
        {
            "description": "Enable pretty-printing for gdb",
            "text": "-enable-pretty-printing",
            "ignoreFailures": true
        },
        {
            "description": "Set Disassembly Flavor to Intel",
            "text": "-gdb-set disassembly-flavor intel",
            "ignoreFailures": true
        }
    ]
  }
  ]
}
```

And how to resolve your include issue I already answered in https://github.com/OpenModelica/OpenModelica/discussions/10325#discussioncomment-5215977
> In the default simulation case OPENMODELICA_XML_FROM_FILE_AT_RUNTIME is defined in the Makefile, so it's no issue that the file is missing.

Define `OPENMODELICA_XML_FROM_FILE_AT_RUNTIME ` and the compiler won't search for the missing C files. They are not generated on purpose.

I guess this is how far we can go in this discussion/question.

**Reply by mtraboulsi1** [reply]
@AnHeuermann Thank you for your reply. I added "Step_init.c" and "Step_Info.c" in Makefile (in CFILES) but the errors are not resolved.
Do I need to add or define other things in Makefile ? Or I have to define them in another place in makefile ?
Thank you again for your help !
![Capture](https://user-images.githubusercontent.com/127110869/226380213-89fada0b-6ad3-41e9-a8f5-9309ed458793.PNG)

**Reply by AnHeuermann** [reply]
Don't add `Step_init.c` and `Step_info.c` to the C files. They are not there, or are they?
What is the error message you'll get when compiling `Step` with the makefile? It should simply work (assuming you have all OpenModelica tools in your path) and have a working OpenModelica installation.

**Reply by mtraboulsi1** [reply]
@AnHeuermann I am in the right directory where the makefile is but I have this error
![Capture1](https://user-images.githubusercontent.com/127110869/226399114-5b32f2b7-0f2e-4cec-80f8-dbf0ce5797eb.PNG)
 message when compiling.

**Reply by AnHeuermann** [reply]
Check the [man page](https://man7.org/linux/man-pages/man1/make.1.html) of make to see how to specify a makefile that is not named `Makefile`. That is why flag `-f` is added to the make command, but you need to replace `path/to/Makefile` with the path to the makefile.

Also you are very likely not using the correct terminal. Check the [OMDev](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/README.Windows.md#1-omdev-package) and use the MSYS2 MINGW64 shell.

**Reply by mtraboulsi1** [reply]
@AnHeuermann While compiling "makefile" I received this error. Do you have any idea how I can solve this problem.
Thank you again for your help. 
![Capture2](https://user-images.githubusercontent.com/127110869/226554802-e6e829a8-55cc-47da-99fb-179bed577ea0.PNG)

**Reply by mtraboulsi1** [reply]
@AnHeuermann I tried to compile Makefile with MSYS2 Mingw64 shell but I got the same error.
Can you help me please?
![Capture3](https://user-images.githubusercontent.com/127110869/226570014-3093a0ce-1222-4e06-b4ab-5ed62f342419.PNG)

**Reply by AnHeuermann** [reply]
Please add the full log of the command you are running. Otherwise we have to guess what you are doing

How did you start the mingw64 shell? Where is it installed?
It looks like you are still not using the msys from OMDev, but installed one yourself. In that case you are missing some dependencies (hdf5 and szip in this case). Please follow the instructions in https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/README.Windows.md#12-install-omdev to use the MSYS shell from the OMDev package. It has all dependencies to debug and compile OpenModelica simulation executables.

You should have library `libhdf5.a` and `libhdf5.dl.a` somewhere in your msys environment. When you use the msys shell provided by OMDev it is in `C:\OMDev\tools\msys\mingw64\lib`. If you use the OpenModelica Command Prompt (OMEdit: Tools->OpenModelica Command Prompt) it is in `C:\Program Files\OpenModelica1.21.0-dev-64bit\tools\msys\mingw64\lib` (or similar).
You can check `LDFLAGS` from `Step.makefile` to see in which directories the linker will search for `hdf5` or `szip`. But if you have a working OpenModelica no changes should be needed.

Also having spaces in your path is usually a bad idea, but should work in this case.

**Reply by mtraboulsi1** [reply]
@AnHeuermann This is the full log of the command I am running. Then, I followed the instructions to use MSYS2 Mingw64 shell from the OMDev package and then I launched the mingw64 directly from OMDev. (C:\OMDev\tools\msys\mingw64.exe).
![1](https://user-images.githubusercontent.com/127110869/226664588-b49f9281-edbc-4dd9-a022-d53e31e59731.PNG)
![2](https://user-images.githubusercontent.com/127110869/226664593-1284fd75-502d-4fcf-82b9-dfec49de1e5c.PNG)
![3](https://user-images.githubusercontent.com/127110869/226664594-aaeb062e-d829-4659-a3bf-a35d0ad7dfee.PNG)
![4](https://user-images.githubusercontent.com/127110869/226664596-75b335e9-8c6e-48ed-b98e-bc6366a08247.PNG)
![5](https://user-images.githubusercontent.com/127110869/226664597-9ad12af2-c1ca-4152-8898-100a90935ff1.PNG)
![6](https://user-images.githubusercontent.com/127110869/226664604-c2acf098-76ee-4e91-b148-5383d91199c7.PNG)
![Mingw64](https://user-images.githubusercontent.com/127110869/226664607-ff14a1cb-7e80-427b-9721-fffd1c114138.PNG)

**Reply by AnHeuermann** [reply]
Please add text files instead of screenshots for logs (easier to read/search/copy text)
You can redirect the output into a file by calling e.g.

```bash
# make .f Step.makefile &> log.txt
```

But I can't see any error messages, so it looks like it worked 🎉 
You can run the executable with

```bash
# ./Step.exe
LOG_SUCCESS       | info    | The initialization finished successfully without homotopy method.
LOG_SUCCESS       | info    | The simulation finished successfully.
```

**Reply by mtraboulsi1** [reply]
@AnHeuermann I can run the executable with ".\Step.bat" but i can't with ".\Step.exe" (Step.exe is in the right direction). Another question how I can see the code output result ?
![Step](https://user-images.githubusercontent.com/127110869/226684237-3270709d-c0c4-48ed-87ec-6ae00065a47b.PNG)

**Reply by mtraboulsi1** [reply]
@AnHeuermann Thank you for your help all this time to run the code and for all the information. Then I have a question: is it possible to display the output results in the terminal ? or is it possible to do "printf" to get the result ? (For example the variation of Step as a function of time).

**Reply by AnHeuermann** [reply]
Not really. Use OMEdit if you want to visualize the computed solution from the result file.
There is a [scripting API](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/scripting_api.html?highlight=scripting) available, but its not directly callable from the command line.

**Reply by mtraboulsi1** [reply]
@AnHeuermann Hi, I would greatly appreciate if you could kindly provide me with information regarding the feasibility of compiling the source code "Step.c" without depending on the "Step.exe" executable. I am curious to know if there exists a method to initiate the compilation of "Step.c" directly.

**Reply by AnHeuermann** [reply]
Sorry, I don't understand what your issue is. If you want to compile the generated C code you should use the Makefile.

> compiling the source code "Step.c" without depending on the "Step.exe" executable

What do you mean with this? The executable depends on the C sources, not the other way around.

Anyway, you could ask a new question if this is not related to the original question of this discussion.

**Reply by mtraboulsi1** [reply]
@AnHeuermann I want to compile "Step.c" directly without needing "Step.exe". Because now I can run "Step.exe".

**Reply by AnHeuermann** [reply]
You never need Step.exe to compile Step.c. The exe is the result of the compilation of Step.c (and the other C files).

**Reply by mtraboulsi1** [reply]
@AnHeuermann So when I run this command "./Step" it means I'm compiling the Step.c directly right ?

**Reply by AnHeuermann** [reply]
Sorry, I can't help you. You want to do something relative complicated (using a C debugger on unknown code), but for me it seems you are missing the basics on working with the C language. I would suggest to search for some tutorials on the C language and how to compile (with Make) on a Unix like system / MINGW.

> So when I run this command "./Step" it means I'm compiling the Step.c directly right ?

No. The executable `Step` or `Step.exe` is the simulation run of the Modelica model. So running the executable means to run the simulation.
The compilation of `Step.exe` itself is done by the OpenModelica compiler. It generates the C source files and the Makefile and then uses that Makefile to compile everything into `Step.exe`. Basically it is doing

```bash
make -f Step.makefile
```

**Comment by JohnsonBervigwm** [comment]
One possible solution could be to try and create the file manually by looking at the contents of the "Step_init.xml" file and translating it to C code.
